python/cuml/common/array.py

```diff
@@ -0,0 +1,124 @@
+            self.order = order
+            self.strides = _order_to_strides(order, shape, dtype)
+
+        elif hasattr(data, "__array_interface__") or \
+                hasattr(data, "__cuda_array_interface__"):
+            super(Array, self).__init__(data=data, owner=owner)
+            self.shape = data.shape
+            self.dtype = np.dtype(data.dtype)
+            self.strides = data.strides
+            self.order = _strides_to_order(data.strides, data.dtype)
+
+        else:
+            raise TypeError("Unrecognized data type.")
+
+    def __getitem__(self, slice):
+        return Array(data=cp.asarray(self).__getitem__(slice))
+
+    def __setitem__(self, slice, value):
+        cp.asarray(self).__setitem__(slice, value)
+
+    def to_output(self, output_type='cupy'):
+        if output_type == 'cupy':
+            return cp.asarray(self)
+
+        elif output_type == 'numba':
+            return cuda.as_cuda_array(self)
+
+        elif output_type == 'numpy':
+            return np.array(cp.asnumpy(cp.asarray(self)), order=self.order)
+
+        elif output_type == 'dataframe':
+            if self.dtype not in [np.uint8, np.uint16, np.uint32,
+                                  np.uint64, np.float16]:
+                mat = cuda.as_cuda_array(self)
+                if len(mat.shape) == 1:
+                    mat = mat.reshape(mat.shape[0], 1)
+                return DataFrame.from_gpu_matrix(mat)
+            else:
+                raise ValueError('cuDF unsupported Array dtype')
+
+        elif output_type == 'series':
+            if len(self.shape) == 1:
+                if self.dtype not in [np.uint8, np.uint16, np.uint32,
+                                      np.uint64, np.float16]:
+                    return Series(self, dtype=self.dtype)
+                else:
+                    raise ValueError('cuDF unsupported Array dtype')
+            elif self.shape[0] > 1 and self.shape[1] > 1:
+                raise ValueError('Only single dimensional arrays can be \
+                                 transformed to cuDF Series. ')
+            else:
+                if self.dtype not in [np.uint8, np.uint16, np.uint32,
+                                      np.uint64, np.float16]:
+                    return Series(self, dtype=self.dtype)
+                else:
+                    raise ValueError('cuDF unsupported Array dtype')
+
+    def __reduce__(self):
+        return self.__class__, (self.to_output('numpy'),)
+
+    @classmethod
+    def empty(cls, shape, dtype, order='F'):
+        size, _ = _get_size_from_shape(shape, dtype)
+        dbuf = DeviceBuffer(size=size)
+        return Array(data=dbuf, shape=shape, dtype=dtype, order=order)
+
+    @classmethod
+    def zeros(cls, shape, dtype='float32', order='F'):
+        return Array(rmm_cupy_ary(cp.zeros, shape, order=order,
+                                  dtype=dtype))
+
+    @property
+    def __cuda_array_interface__(self):
```

Comment:
Will this play correctly with `__array_function__` support from cupy? E.g. will `np.unique(my_cuml_Array)` actually call cp.unique as with cupy? If not, maybe we can plug into cupy's existing functions there somehow?
Cupy also provides the nice `get_array_module()` interface (which just returns cp for cupy arrays and np for numpy arrays) - not widely used but a good step towards abstraction.

Reply:
Not guaranteed to work and I don't think we want to make it work. I had a prior more full featured verion (that worked for that), but one of the main concerns developing this is that we do not want it to be *yet another* array library, and in our codebase we would not do `np.unique(cuml_ary)`, we would do `cp.unique(cuml_ary)`. If we really want to use the numpy function with cupy's `__array_function__` support  we can use cupy's support itself with `np.unique(cp.asarray(cuml_ary))` which does not seem like too much of a burden since this is meant for developer and not user consumption. What do you think?

Reply:
I should add that this would not require big additions to the code if I'm not mistaken (pending more testing to be sure that statement is true). My argument is more of I think we want to keep the shim cuml Array as minimal as possible and add the features we need. If/once we need `__array_function__` support we can easily add it. This is why I keep calling it a shim instead of a proper array